In [1]:
import numpy as np

In [2]:
#create empty sigmoid function
def sigmoid(x):
    return 1/(1 + np.exp(-x))

In [3]:
sigmoid(-234)

2.371871925555801e-102

In [4]:
#generate training example
train_data = np.array([[0,0,1],[1,1,1],[1,0,1],[0,1,1]])

train_output = np.array([[0,1,1,0]]).T

In [5]:
train_output

array([[0],
       [1],
       [1],
       [0]])

In [6]:
train_data

array([[0, 0, 1],
       [1, 1, 1],
       [1, 0, 1],
       [0, 1, 1]])

In [7]:
train_output

array([[0],
       [1],
       [1],
       [0]])

In [8]:
#to check size 
train_data.size

12

In [9]:
#to check dimension
train_data.ndim

2

In [10]:
#to check shape
train_data.shape

(4, 3)

In [11]:
#generate weights
#formula to calc weights = summation of x and w (x is input and w is weights)
np.random.seed(1)
weights = 2 * np.random.random((3,1))-1
weights

array([[-0.16595599],
       [ 0.44064899],
       [-0.99977125]])

In [12]:
for i in range(1000):
    outputs = sigmoid(np.dot(train_data, weights))
print(outputs)

[[0.2689864 ]
 [0.3262757 ]
 [0.23762817]
 [0.36375058]]


weights  = 0(-0.16595599) + 0( 0.44064899) + 0(-0.99977125) = -0.9999

sigmoid = 1/1+e^-(0.999) = 0.2689864

calculation of error
error = actualoutput - output

Formula to calculate SigmoidCurveGradient
SigmoidCurveGradient(ouput)=output.(1-output)

In [13]:
def SCG(x):
    return x * (1-x)

In [14]:
for i in range(50000):
    inputs = train_data

    outputs = sigmoid(np.dot(inputs, weights))

    error = train_output - outputs

    adjusted_weights = error * SCG(outputs)

    weights += np.dot(inputs.T, adjusted_weights)

print(weights)
print("------------------")
print(outputs)

[[11.30926129]
 [-0.20509237]
 [-5.45001623]]
------------------
[[0.0042779 ]
 [0.99650925]
 [0.99715469]
 [0.00348742]]
